### PR TITLE
Prep work to migrate to PHPUnit 9.x

### PR DESCRIPTION
**Pull request by sanmai**
**This is a fully backward-compatible change.** 

The general idea is to make the package testable under PHP 5.6 - 7.x - 8.0. This requires using both PHPUnit 5.x and 9.x, because only the last works correctly under PHP 8 (as is - collects coverage report).

But PHPUnit 9 has quite a lot of legacy methods removed, and PHPUnit 8 has these methods but with `void` return type. Therefore implementing these methods straight away is a no-go. This can be [with a version-dependent trait](https://github.com/symfony/phpunit-bridge/blob/5.x/Legacy/PolyfillAssertTrait.php), but for a small handful of functions used here it easier to do this by overloading them from `__call` and `__callStatic`. This is where `DeprecatedTestMethods` trait comes into play. 

And there are also `setUp(): void` etc. which are **not** taken care in this PR.

- There's also above-mentioned `PolyfillAssertTrait` from `symfony/phpunit-bridge`, but it is tagged `@internal` so it's a risky game to use it.
- And there's also [phpunitgoodpractices/polyfill](https://packagist.org/packages/phpunitgoodpractices/polyfill), but it should be trivial to switch to any of those down the line.

---
Extracted from #5262 (it handles  `setUp(): void` among other things)

### CODE REVIEW

File: tests/AutoReview/TravisTest.php
Comment (sanmai) [inline tests/AutoReview/TravisTest.php:141]: On PHP 8 it'll be `[7, 7.1, 7.2...]`, and casting first to float will result in just `7` not `7.0`.

File: phpstan.neon
Comment (GrahamCampbell) [inline phpstan.neon:17]: Why do we need this?

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: phpstan.neon
Comment (sanmai) [reply phpstan.neon:17]: Because PHPStan will fail the build without it. And _why_ it thinks we may call any `assert*` instance methods (where we know they're all `static`) is beyond me.

Comment (GrahamCampbell) [reply phpstan.neon:17]: Not a bug. It'll be caused by the `__call` method you have?

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: phpstan.neon
Comment (sanmai) [reply phpstan.neon:17]: If you take a look at the header of `DeprecatedTestMethods` you'll notice that all `assert*` methods declared there are `static`. Here's [a similar issue](https://github.com/phpstan/phpstan/issues/1557), and a proposed workaround there is the same used here.

Comment (sanmai) [reply phpstan.neon:17]: We need `__call` for `expectException*` methods. They're indeed instance methods. PHPStan has nothing against them.

Comment (sanmai) [reply phpstan.neon:17]: Just for the record, removing `__call` does nothing to the issue. These are still there:

```
 ------ ----------------------------------------------------------------------------- 
  Line   tests/RuleSet/Sets/AbstractSetTest.php                                       
 ------ ----------------------------------------------------------------------------- 
  66     Static call to instance method PhpCsFixer\Tests\TestCase::assertIsString().  
 ------ ----------------------------------------------------------------------------- 

 ------ --------------------------------------------------------------------------------------------- 
  Line   tests/Test/AbstractFixerTestCase.php                                                         
 ------ --------------------------------------------------------------------------------------------- 
  484    Static call to instance method PhpCsFixer\Tests\TestCase::assertIsString().                  
  485    Static call to instance method PhpCsFixer\Tests\TestCase::assertMatchesRegularExpression().  
  486    Static call to instance method PhpCsFixer\Tests\TestCase::assertStringNotContainsString().   
 ------ --------------------------------------------------------------------------------------------- 

 ------ ---------------------------------------------------------------------------- 
  Line   tests/Tokenizer/TokensTest.php                                              
 ------ ---------------------------------------------------------------------------- 
  1411   Static call to instance method PhpCsFixer\Tests\TestCase::assertIsArray().  
  1418   Static call to instance method PhpCsFixer\Tests\TestCase::assertIsArray().  
 ------ ---------------------------------------------------------------------------- 
```
You can [check with `TokensTest.php`](https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/3f1b03d123898e1a5a0b9fba3c6f2cb7d4fc39c5/tests/Tokenizer/TokensTest.php#L1411) just for an example, to see that PHPStan is telling a pure BS here.

Comment (GrahamCampbell) [reply phpstan.neon:17]: Hmmm. Can we come up with a minimal example, so we can report the issue to them?

Comment (sanmai) [reply phpstan.neon:17]: Vanilla PHPStan has nothing like that: https://phpstan.org/r/16b02a02-a9b5-4b65-9aa2-30a0e4ffd586

Comment (sanmai) [reply phpstan.neon:17]: Commenting out `phpstan/phpstan-phpunit/extension.neon` alleviates the issue. If you wish to report the issue to `phpstan/phpstan-phpunit`, please do. I'm not going to have the time to follow them up.

File: tests/Console/Command/FixCommandTest.php
Comment (GrahamCampbell) [inline tests/Console/Command/FixCommandTest.php:34]: should be `static::`

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: tests/Console/Command/FixCommandTest.php
Comment (sanmai) [reply tests/Console/Command/FixCommandTest.php:34]: Why so? [It's an instance method.](https://github.com/sebastianbergmann/phpunit/blob/eafcb3a19339f99232354dbf949bd2f890f2491e/src/Framework/TestCase.php#L619-L622)

Comment (sanmai) [reply tests/Console/Command/FixCommandTest.php:34]: Why this pointless nitpicking, mind you?

**Comment by sanmai** [comment]
@keradus should it be more appropriate to use `phpunitgoodpractices/polyfill` here instead of a purpose-build short polyfill?

**Comment by SpacePossum** [comment]
utilizing `phpunitgoodpractices/polyfill` would be great as more people found benefit from your work : )

(also we would have less maintenance in this project so I'm biased ;) )

**Comment by sanmai** [comment]
Good point. Give me a minute to sort this out.

**Comment by sanmai** [comment]
Should I add `phpunitgoodpractices/polyfill` as a direct dependency?

**Comment by keradus** [comment]
:+1: 
feel free to raise PRs to that polyfill repo too, don't hesitate to reach out to me over https://gitter.im/keradus

**Comment by sanmai** [comment]
@keradus DM'ing you there!

**Comment by sanmai** [comment]
First PR for `phpunitgoodpractices/polyfill` is go https://github.com/PHPUnitGoodPractices/polyfill/pull/16

Once there's a CI, I'll add other methods. 

Meanwhile this is ready for merge. Migrating to  `phpunitgoodpractices/polyfill` will be as simple as removing one file and changing one line.

File: .php_cs.dist
Comment (keradus) [inline .php_cs.dist:26]: nice :+1:

File: tests/Smoke/CiIntegrationTest.php
Comment (keradus) [inline tests/Smoke/CiIntegrationTest.php:160]: just to ensure - this change is done manually or automatically by one of the PHP CS Fixer rule ?

File: tests/Test/AbstractFixerTestCase.php
Comment (keradus) [inline tests/Test/AbstractFixerTestCase.php:146]: just to ensure - this change is done manually or automatically by one of the PHP CS Fixer rule ?

File: tests/Smoke/CiIntegrationTest.php
Comment (sanmai) [reply tests/Smoke/CiIntegrationTest.php:160]: Not it wasn't. It should have been, but it wasn't. 

There's no such `@PHPUnitXXMigration` rule-set that would do this, as far as I can tell.

File: tests/Test/AbstractFixerTestCase.php
Comment (sanmai) [reply tests/Test/AbstractFixerTestCase.php:146]: This is done by `php_unit_dedicate_assert_internal_type`

File: tests/Smoke/CiIntegrationTest.php
Comment (keradus) [reply tests/Smoke/CiIntegrationTest.php:160]: can you raise a feature request for it, please?

**Comment by GrahamCampbell** [comment]
What is the reason for the re-target?

**Comment by GrahamCampbell** [comment]
Oh, I see. The last 2.15.x release has just happened?

File: tests/Console/Command/FixCommandTest.php
Comment (GrahamCampbell) [reply tests/Console/Command/FixCommandTest.php:34]: The review process on this repo is rigorous. We are all grateful for contributions from everyone, however it is also in everyone's interest to make sure the code is of good quality, and works properly.

**Comment by GrahamCampbell** [comment]
So, what is the status with `phpunitgoodpractices/polyfill`? I think this is the only outstanding thing blocking merge now. The rest looks great, from what I can see.

**Comment by sanmai** [comment]
@GrahamCampbell what we're doing there is fairly tricky, and, considering that @keradus is, understandably, looking for an uncompromising quality, at these circumstances I'd like to see it getting as thorough review as possible. Therefore, we have to wait for it to happen.

File: tests/Console/Command/FixCommandTest.php
Comment (keradus) [reply tests/Console/Command/FixCommandTest.php:34]: I do not understand the request as well.
it was call on instance of object, it is still like that. the method is not static. why to change it?

as it's keeping the current status of repo, I'm happy to merge the PR without suggested change and have the change discussed separately (and maybe even merged, as separated PR?)

Comment (GrahamCampbell) [reply tests/Console/Command/FixCommandTest.php:34]: Yeh, this review comment can be marked as done.

**Review by keradus — APPROVED** [review]
nice one ;)

File: tests/Smoke/CiIntegrationTest.php
Comment (sanmai) [reply tests/Smoke/CiIntegrationTest.php:160]: Done: #5294

**Comment by sanmai** [comment]
Do I need to squash-rebase it? It's not a big deal, but squash merging will be easier.

**Comment by keradus** [comment]
squash doesn't matter.
i only await for CI to confirm it works on all supported combinations of phpunit

**Review by GrahamCampbell — APPROVED** [review]

**Comment by GrahamCampbell** [comment]
This can be (squash) merged now, I think?

**Comment by keradus** [comment]
easy @GrahamCampbell , it will be merged when it will be merged. no need for pinging, especially when you see movement around related changes

**Comment by GrahamCampbell** [comment]
I wasn't rushing. I was just re-approving after the changes you just made. :)

**Comment by keradus** [comment]
Thank you @sanmai.

**Pull request merged**
